webserver/views/api_compat.py

```diff
@@ -0,0 +1,352 @@
+import time
+import json
+import re
+from collections import defaultdict
+from yattag import Doc
+import yattag
+from flask import Blueprint, request, render_template
+from flask_login import login_required, current_user
+from webserver.external import messybrainz
+from webserver.rate_limiter import ratelimit
+from webserver.errors import InvalidAPIUsage
+import xmltodict
+from api_tools import insert_payload
+from db.lastfm_user import User
+from db.lastfm_session import Session
+from db.lastfm_token import Token
+import calendar
+from datetime import datetime
+
+api_bp = Blueprint('api_compat', __name__)
+
+
+@api_bp.route('/api/auth/', methods=['GET'])
+@ratelimit()
+@login_required
+def api_auth():
+    """ Renders the token activation page.
+    """
+    token = request.args['token']
+    return render_template(
+        "user/auth.html",
+        user_id=current_user.musicbrainz_id,
+        token=token
+    )
+
+
+@api_bp.route('/api/auth/', methods=['POST'])
+@ratelimit()
+@login_required
+def api_auth_approve():
+    """ Authenticate the user token provided.
+    """
+    user = User.load_by_name(current_user.musicbrainz_id)
+    token = Token.load(request.form['token'])
+    if not token:
+        return render_template(
+            "user/auth.html",
+            user_id=current_user.musicbrainz_id,
+            msg="Either this token is already used or invalid. Please try again."
+        )
+
+    if token.api_key != user.api_key:
+        return render_template(
+            "user/auth.html",
+            user_id=current_user.musicbrainz_id,
+            msg="This token does not belong to this account. Please login in the correct account."
+        )
+    if token.user:
+        return render_template(
+            "user/auth.html",
+            user_id=current_user.musicbrainz_id,
+            msg="This token is already approved. Please check the token and try again."
+        )
+    if token.has_expired():
+        return render_template(
+            "user/auth.html",
+            user_id=current_user.musicbrainz_id,
+            msg="This token has expired. Please create a new token and try again."
+        )
+    token.approve(user.name)
+    return render_template(
+        "user/auth.html",
+        user_id=current_user.musicbrainz_id,
+        msg="Token %s approved for user %s, press continue in client." % (token.token, current_user.musicbrainz_id)
+    )
+
+
+@api_bp.route('/2.0/', methods=['POST'])
+@ratelimit()
+def api_post():
+    """ Receives the POST-API calls and redirects them to appropriate methods.
+    """
+    method = request.form['method'].lower()
+    def invalid_method_error(*args):
+        raise InvalidAPIUsage(3, output_format=request.args.get('format', "xml"))   # Invalid Method
+
+    return {
+        'track.updatenowplaying': record_listens,
+        'track.scrobble': record_listens,
+        'auth.getsession': get_session,
+        'auth.gettoken': get_token,
+        'user.getinfo': user_info
```

Comment:
This is a novel way of choosing which helper function to call based on the `method` parameter, but it took me a few seconds to work out what the code did. There is a lot happening here (return statement, dict get, default value, function execution).

There's a famous programming quote which is worth keeping in mind here:

> “Debugging is twice as hard as writing the code in the first place. Therefore, if you write the code as cleverly as possible, you are, by definition, not smart enough to debug it.”
> https://blogs.msdn.microsoft.com/alfredth/2009/01/08/are-you-smart-enough-to-debug-your-own-code/

I would prefer this to be a series of `if` statements, to make it easier to quickly understand how it works.